packages/concerto-tools/lib/codegen/fromcto/protobuf/protobufvisitor.js

```diff
@@ -0,0 +1,440 @@
+     * @return {Object} a package name compliant with Proto 3
+     * @public
+     */
+    concertoNamespaceToProto3SafePackageName(concertoNamespace) {
+        // Proto3 needs the namespace to have a standard Java-like format, so the "@" and the dots in the version need to be replaces with underscores.
+        return `${concertoNamespace.split('@')[0]}.v${concertoNamespace.split('@')[1].replace(/\./ig, '_')}`;
+    }
+
+    /**
+     * Transform a Concerto meta property into a Proto3 field rule.
+     * @param {Object} field - the Concerto meta property
+     * @return {Object} the Proto3 field rule
+     * @public
+     */
+    concertoToProto3FieldRule(field) {
+        // An array (repeated) in Proto3 is implicitly optional.
+        if (field.isArray()) {
+            return 'repeated';
+        } else if (field.isOptional()) {
+            return 'optional';
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Transform a Concerto primitive type into a Proto3 one.
+     * @param {Object} field - the Concerto primitive type
+     * @return {Object} the Proto3 primitive type
+     * @public
+     */
+    concertoToProto3PrimitiveType(field) {
+        switch (field.getType()) {
+        case 'String':
+            return 'string';
+        case 'Double':
+            return 'double';
+        case 'Integer':
+            return 'sint64';
+        case 'Long':
+            return 'sint64';
+        case 'DateTime':
+            return 'google.protobuf.Timestamp';
+        case 'Boolean':
+            return 'bool';
+        }
+    }
+
+    /**
+     * Transform a Concerto class or enum type into a Proto3 message or enum one.
+     * @param {Object} field - the Concerto class or enum type
+     * @return {Object} the Proto3 message or enum type
+     * @public
+     */
+    concertoToProto3MessageOrEnumType(field) {
+        return this.doesClassHaveChildren(
+            field.getType(), field.parent.modelFile.declarations
+        )
+            ?  `_Children_of_class_${field.getType()}`
+            : field.getType();
+    }
+
+    /**
+     * Transform Concerto class imports to Proto3 import line strings.
+     * @param {Object[]} imports - the imports of a Concerto class
+     * @return {string[]} an array of import line strings
+     * @public
+     */
+    createImportLineStrings(imports) {
+        return imports
+            .filter(importObject => importObject.namespace.split('@')[0] !== 'concerto')
+            .map(importObject => `${this.concertoNamespaceToProto3SafePackageName(importObject.namespace)}.proto`);
+    }
+
+    /**
+     * Get the names of the children of a class.
+     * @param {string} className - the name of the class
+     * @param {Object[]} declarations - the declarations in scope
+     * @return {string[]} an array of the names of the children of the classes
+     * @public
+     */
+    getChildrenOfClass(className, declarations) {
```

Comment:
I think you should maybe use:
https://github.com/accordproject/concerto/blob/93659e3b59370729ac62c466009f88a56fc32758/packages/concerto-core/lib/introspect/classdeclaration.js#L456

This will recursively return all specialisations of a given class (including the class itself).

Have you tested the code with multiple levels of inheritance?